packages/autocomplete-core/src/utils/createCancelablePromise.ts

```diff
@@ -0,0 +1,162 @@
+      return createCancelable(
+        promise.then(
+          createCallback(onfulfilled, state, promise),
+          createCallback(onrejected, state, promise)
+        ),
+        state
+      );
+    },
+    catch(onrejected) {
+      return createCancelable(
+        promise.catch(createCallback(onrejected, state, promise)),
+        state
+      );
+    },
+    finally(onfinally) {
+      if (onfinally) {
+        state.onCancelList.push(onfinally);
+      }
+
+      return createCancelable<TValue>(
+        promise.finally(
+          createCallback(
+            onfinally &&
+              (() => {
+                state.onCancelList = [];
+
+                return onfinally();
+              }),
+            state,
+            promise
+          )
+        ),
+        state
+      );
+    },
+    cancel() {
+      state.isCanceled = true;
+      const callbacks = state.onCancelList;
+      state.onCancelList = [];
+
+      callbacks.forEach((callback) => {
+        callback();
+      });
+    },
+    isCanceled() {
+      return state.isCanceled === true;
+    },
+  };
+}
+
+export type CancelablePromise<TValue> = {
+  then<TResultFulfilled = TValue, TResultRejected = never>(
+    onfulfilled?:
+      | ((
+          value: TValue
+        ) =>
+          | TResultFulfilled
+          | PromiseLike<TResultFulfilled>
+          | CancelablePromise<TResultFulfilled>)
+      | undefined
+      | null,
+    onrejected?:
+      | ((
+          reason: any
+        ) =>
+          | TResultRejected
+          | PromiseLike<TResultRejected>
+          | CancelablePromise<TResultRejected>)
+      | undefined
+      | null
+  ): CancelablePromise<TResultFulfilled | TResultRejected>;
+  catch<TResult = never>(
+    onrejected?:
+      | ((
+          reason: any
+        ) => TResult | PromiseLike<TResult> | CancelablePromise<TResult>)
+      | undefined
+      | null
+  ): CancelablePromise<TValue | TResult>;
+  finally(
+    onfinally?: (() => void) | undefined | null
+  ): CancelablePromise<TValue>;
+  cancel(): void;
+  isCanceled(): boolean;
+};
+
+export function createCancelablePromise<TValue>(
+  executor: PromiseExecutor<TValue>
+): CancelablePromise<TValue> {
+  return createInternalCancelablePromise({
+    initialState: createInitialState(),
+    promise: new Promise<TValue>((resolve, reject) => {
+      return executor(resolve, reject);
+    }),
+  });
+}
+
+createCancelablePromise.resolve = <TValue>(
+  value?: TValue | PromiseLike<TValue> | CancelablePromise<TValue>
+) => cancelable(Promise.resolve(value));
+
+createCancelablePromise.reject = (reason?: any) =>
+  cancelable(Promise.reject(reason));
+
+export function cancelable<TValue>(promise: Promise<TValue>) {
+  return createCancelable(promise);
+}
+
+function createCancelable<TValue>(
+  promise: Promise<TValue>,
+  initialState: CancelablePromiseState = createInitialState()
+) {
+  return createInternalCancelablePromise<TValue>({
+    promise,
+    initialState,
+  });
+}
```

Comment:
We don't need this default if we pass it here, which makes the code more explicit. Wdyt?

```suggestion
export function cancelable<TValue>(promise: Promise<TValue>) {
  return createCancelable(promise, createInitialState());
}

function createCancelable<TValue>(
  promise: Promise<TValue>,
  initialState: CancelablePromiseState
) {
  return createInternalCancelablePromise<TValue>({
    promise,
    initialState,
  });
}
```

Reply:
As the item gets recreated on every function call anyway, this is even simpler imo:

```suggestion
export function cancelable<TValue>(promise: Promise<TValue>) {
  return createInternalCancelablePromise<TValue>({
    promise,
    initialState: createInitialState(),
  });
}
```

maybe even inlined createInitialState?

Reply:
```suggestion
export function cancelable<TValue>(promise: Promise<TValue>) {
  return createInternalCancelablePromise<TValue>({
    promise,
    initialState: { isCanceled: false, onCancelList: [] },
  });
}
```

this is my preferred form, it avoids more internal functions

Reply:
@Haroenv We need to be able to pass an initial state when returning cancelable promises in `then`, `catch` and `finally`, that's why `createCancelable` exists. But @francoischalifour's refactor would work and we shave some more bytes by stripping the default value.